docs(draft): new version of TRG 4.02 - Base Images (#594)
eclipse-tractusx/eclipse-tractusx.github.io: docs/release/trg-0/trg-4-02.md

---
title: TRG 4.02 - Base Images
---

| Status | Created     | Post-History                                                                    |
|--------|-------------|---------------------------------------------------------------------------------|
| Draft  | 15-Jan-2024 | Add Base Image Security                                                         |
| Active | 06-Dec-2023 | Add advantages of Dockerhub images                                              |
| Draft  | 04-May-2023 | Remove notice in favor of dedicated TRG; Add mandatory base image for frontends |
| Active | 25-Nov-2022 | Initial release                                                                 |

## Why

As part of our legal due diligence, we need to provide the best information possible about our distributed (published)
Docker images. Similar to our 3rd-party dependency scans and the `DEPENDENCIES` file, Docker images also have to be
scanned and the results published. We want to help you to keep a high standard process, by defining guidelines,
described in this TRG.

## Description

> As Eclipse Tractus-X project, we don't have automated processes for publishing container scan results (yet). This is why

Reviewer: would remove the (yet)

Reviewer: sure no problem, but can we change it now? feels a bit strange for me to write about upcomming maybe plans for the future into a TRG

Author: I won't change this, as it is not the subject of this PR to change other content of the TRG. Even though the change is as little as this one.

or major version:

:::warning Do not use image tag with patch version

```DOCKERFILE
FROM alpine:3.18.5 as BUILDER

COPY . /app/build
WORKDIR /app/build
RUN build

FROM alpine:3.18.5

COPY --from=BUILDER /app/build /app
[...]
```

:::

:::tip Use image tag with minor or major version

```DOCKERFILE
FROM alpine:3 as BUILDER

COPY . /app/build
WORKDIR /app/build
RUN build

FROM alpine:3.18

COPY --from=BUILDER /app/build /app

[...]
```

:::

Using minor image tags instead of patch level image tags avoid to touch the applications `DOCKERFILE` to fix a
potential problem inside the base image. With the next release of the Eclipse-Tractus-X application/component an updated
base image will be used for the application, if available.

### Base Image Security

We trust in the maintenance and security mechanisms of the base images which are aligned with this TRG (see [table
above](#base-images-to-use)). Don't try to fix vulnerabilities in base images. Instead, we focus ourselves to keep our
code free of vulnerabilities.

You shall not change the content of the base image to fix vulnerabilities of the base image using package
manager update mechanisms inside a `DOCKERFILE`:

```DOCKERFILE

> FROM alpine:3.18.2

Reviewer: as a good example would also remove the patch version information

Author: Get you point, but this is an example using patchlevel image tag **and** updating/chaning the base image using package manager, so I would stay with this example.

Reviewer: Got your point, my intention was ... do you really want to separate the "bad" examples, one for patchlevel "Do not use image tag with patch version" and another example for "updating/chaning the base image" **and** "example using patchlevel image"?

Maybe you can combine it in a bigger example in the `warning` section and this would also keep it simple, when decide to one TRG update instead of separating it.

If the language or runtime environment of your product is not listed above, feel free
to [start a discussion](https://github.com/eclipse-tractusx/sig-infra/discussions) and propose your preferred container
images as base image.

:::info
As stated in the description above, base image usage is particularly aligned for container images, that we distribute by
publishing them on DockerHub. In case you are using Docker images for build or testing purposes (for example _pandoc_ or
_cypress_, etc.) and you do not publish the images, you can use other publicly available image, as long as the tools are
open source license compliant.

For automated TRG checks, you can skip base image checks on Dockerfiles by declaring it in the `.tractusx` metadata
files. Details can be found in
the [metadata file documentation](https://github.com/eclipse-tractusx/tractusx-quality-checks/blob/main/docs/metadata_file.md)
:::

### Use minor or major Image Tag

Do not use an image tag pointing to a patch version of the image. Instead, use an image tag containing either minor or
major version.

To keep application images compliant, up-to-date and secure it is mandatory to use image tags which allows to rebuild

> the applications container image without touching the `DOCKERFILE` to pull an updated version of the base image (e.g. to

Reviewer 2: `Dockerfile` is usually not written in caps.

Author: Fixed with latest commit.

fix a potential problem inside the base image.

:::warning Do not use image tag with patch version

```DOCKERFILE
FROM alpine:3.18.5 as BUILDER

COPY . /app/build
WORKDIR /app/build
RUN build

FROM alpine:3.18.5

COPY --from=BUILDER /app/build /app
[...]
```

:::

:::tip Use image tag with minor or major version

```DOCKERFILE
FROM alpine:3 as BUILDER

COPY . /app/build
WORKDIR /app/build
RUN build

FROM alpine:3.18

COPY --from=BUILDER /app/build /app

[...]
```

:::

The next time the applications container image is created, an updated base image will be used, if available.

### Use Base Image as-is

> It is recommended to use the base image **as-is** and to not modify or update it.

Reviewer: ```suggestion
Utilize the base image as provided **without any** modifications.
```

Do not use package manager update mechanisms to update the runtime of the base image (e.g. to fix security
vulnerabilities). Updating or modifying the base image might pull in a new version of a library which is no longer under
the correct Intellectual Property and/or license.

Same applies for installing additional tools, libraries, commands, which modifies the base image. If possible, select a
base image for your application that contains all dependent tools.

:::warning Do not update or install additional packages

```DOCKERFILE
# Call for feedback: Does this example make sense at all or 
# should it be removed?
FROM alpine:3.18.2

COPY some/binary /app/
RUN apk update && apk upgrade \ # fix CVE-2023-123
    apk add someTool            # install pkg "someTool"
```

:::

In the example a patch level image tag of alpine is used, which is not recommended, and an update/upgrade is done to get
the latest versions of all installed packages installed. Instead, use a minor image tag (e.g. `alpine:3.18`) as
recommended in [_Use minor or major Image Tag_](#use-minor-or-major-image-tag).

> An exception from modifying the base image might be a build stage in the `DOCKERFILE`, as long as the resulting image

Reviewer 2: Would again write it like it is usually used: `Dockerfile`

Author: Fixed with latest commit.
